### PR TITLE
fix(107): hyphen-safe legacy v2 SDK session ID parser

### PR DESCRIPTION
**Pull request by isaiahb**
## Summary

The v2 SDK CONNECTION_INIT / RECONNECT session ID format is `<userId>-<packageName>`. The cloud parsed it with `indexOf("-")` (originally `split("-")[0]`), which silently broke any user whose email contained a hyphen — every WS upgrade was rejected with `1008 "Session not found"`, and v2 SDK apps got stuck in retry loops.

Example: `a-b@example.com` running `some-app-name` produces session ID `a-b@example.com-some-app-name`. The pre-fix parser splits at the first hyphen (position 1) and extracts `userId = "a"`. `UserSession.getById("a")` returns undefined, the cloud closes the WS with `1008`, the SDK retries, the cycle repeats. The fingerprint of this bug firing in error logs is a single-character or otherwise truncated userId where an email was expected.

**Anyone with a hyphen anywhere in their email — local-part or domain — has been silently broken since this code shipped.**

## The fix

`cloud/packages/cloud/src/services/websocket/legacy-session-id.ts` (new module):

- Enumerate every split position of the input.
- Keep candidates whose userId-portion contains `@`.
- Prefer the longest candidate whose userId matches a live UserSession (injected lookup, O(1) `Map.get` per candidate).
- Fall back to the longest `@`-bearing candidate when no live session matches, so error logs surface a useful userId for diagnostics.

`bun-websocket.ts` simplified:

```diff
+ import { parseLegacySessionId } from "./legacy-session-id";

  function parseUserIdFromLegacySessionId(sessionId?: string): string | undefined {
-   if (!sessionId) return undefined;
-   const separator = sessionId.indexOf("-");
-   if (separator <= 0) return undefined;
-   return sessionId.slice(0, separator);
+   return parseLegacySessionId(sessionId, isActiveUserId)?.userId;
  }
```

The two wrapper signatures are unchanged — all callers continue working without modification.

## Tests (96 pass, 0 fail)

- `legacy-session-id.test.ts` — unit tests covering happy path, hyphen edge cases, malformed input, lookup disambiguation.
- `legacy-session-id.regression.test.ts` — parameterized regression sweep over hyphen-affected and clean email patterns. Each case is run twice: once verifying the new behavior produces the expected userId+packageName, and once verifying the new behavior MATCHES the pre-fix parser on non-hyphenated inputs (no regressions).
- `legacy-session-id.integration.test.ts` — integration tests via `bun-websocket.ts`'s `__forTesting` exports + `spyOn(UserSession, "getById")`, exercising the production wiring end-to-end.

All three files run via `bun test src/services/websocket/legacy-session-id`.

## Backwards compatibility

- v3 SDK: **unaffected**. v3 sends `x-user-id` as an HTTP header at upgrade time; this parser is never called.
- v2 SDK with non-hyphenated email: **no behavior change**. Regression tests verify the new parser produces the same userId+packageName as the pre-fix parser for every non-hyphenated case.
- v2 SDK with hyphenated email: **fixed**. Previously broken, now works.
- Existing sessions stored in MongoDB (`AppSession.sessionId` field, format `` `${userId}-${packageName}` ``): round-trip correctly through the new parser. Tested directly with the format `AppManager.ts:208` constructs.

## Rollout

- This PR targets `staging`, per the staging-first release cycle.
- A separate hotfix PR is open against `main` to get the fix to prod users immediately.
- `.github/workflows/porter-debug.yml` is updated in this PR to add the branch to push triggers, so cloud-debug auto-deploys on every push for soak validation.

## Blast radius

Anyone with `-` anywhere in their email has been unable to use any v2 SDK app since this code shipped. Worth a quick `db.users.countDocuments({ email: /-/ })` post-merge to size the affected population.

## See also

- `cloud/issues/107-legacy-sessionid-hyphen-parse-bug/spike.md` — full investigation
- `cloud/issues/107-legacy-sessionid-hyphen-parse-bug/spec.md` — implementation details + decision log

## Test plan

- [ ] Push to cloud-debug via porter-debug auto-deploy
- [ ] Run staging soak ≥30 min, verify no new errors in BetterStack
- [ ] Cross-check: truncated-userId errors (single-letter userId in `User session not found` logs) drop to zero
- [ ] Manual: send CONNECTION_INIT with sessionId `test-hyphen@example.com-com.example.test` to verify acceptance path
- [ ] Promote staging → main after soak

### CODE REVIEW

File: cloud/packages/cloud/src/services/websocket/legacy-session-id.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        return candidates[i];
+      }
+    }
+  }
+
+  // No live UserSession matches any candidate (or no lookup was provided).
+  // Fall back to the longest `@`-bearing candidate so error logs surface a
+  // useful userId for diagnostics. Caller will hit "Session not found" but
+  // at least the logged userId is recognizable instead of `"s"`.
+  return candidates[candidates.length - 1];
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply cloud/packages/cloud/src/services/websocket/legacy-session-id.ts]: Good catch — valid concern, fixed in the latest push.

The old fallback (`candidates[candidates.length - 1]`) returned the **longest** `@`-bearing candidate, which for `a-b@example.com-some-app-name` yields `userId="a-b@example.com-some-app"` / `packageName="name"` — exactly the truncation you flagged. That same path feeds `validateApiKey(packageName, apiKey)` in the v3-SDK reconnect bootstrap (`bun-websocket.ts:777`), so a wrong package there would reject valid reconnects.

New fallback strategy (`legacy-session-id.ts:99-130`):

1. Among `@`-bearing candidates, pick the **shortest** whose userId ends with a TLD-shaped suffix (`/\.[a-zA-Z]{2,5}$/`). Shortest because the email's TLD is the **first** dot-letter suffix scanning left-to-right; longer candidates have stretched into the package name.
2. If no candidate is TLD-shaped (rare: IP-domain emails, internationalized TLDs), fall back to longest `@`-bearing — preserves diagnostic-userId behavior without regressing further.

The 2–5 letter bound on the TLD is deliberate: covers `.com`, `.org`, `.io`, `.uk`, `.app`, `.glass`, `.world`, etc., while excluding longer dot-letter chunks that are almost certainly subdomain words. Without that bound, `user@sub.hyphen-domain.com-captions` would match `user@sub.hyphen` as "TLD-shaped" (`.hyphen` is 6 letters) and pick the wrong split.

Test coverage tightened to lock it in:
- `legacy-session-id.test.ts`: new explicit cases for shortest-TLD-wins, multiple-TLD-candidates (`user@example.com-pkg.io-other`), `.glass` recognition, and the hyphenated-subdomain guard.
- `legacy-session-id.regression.test.ts`: the fallback sweep now runs over **every** hyphen-affected pattern (was only clean emails). 121 tests pass, 0 fail.
- `legacy-session-id.integration.test.ts`: the "no live session" test now asserts the exact `userId` + `packageName` via `.toBe()` instead of the prior loose `.includes()`.

**Comment by isaiahb** [comment]
@codex review

File: cloud/packages/cloud/src/services/websocket/legacy-session-id.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+// almost certainly subdomain words rather than TLDs (e.g.
+// `user@sub.hyphen-domain.com`'s candidate `user@sub.hyphen` ends with
+// `.hyphen` — 6 letters — which is a subdomain word and would be a
+// false-positive TLD match without this bound).
+//
+// Trade-off: the rarest long TLDs (.museum, .travel, .coffee, ...) are
+// not matched here. Users on those TLDs fall through to the
+// longest-@-bearing fallback below, which preserves the prior behavior —
+// not perfect but not a regression either.
+const TLD_SUFFIX = /\.[a-zA-Z]{2,5}$/;
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply cloud/packages/cloud/src/services/websocket/legacy-session-id.ts:154]: Right call — fixed in the latest push.

The 2–5 letter TLD bound was a real regression for long-gTLD users (`.museum`, `.industries`, `.versicherung`, …) whose pre-fix behavior was correct (no hyphen in domain → first-hyphen split worked). The old final fallback (`candidates[candidates.length - 1]`) then truncated their package.

New final fallback (`legacy-session-id.ts:99-117`): when no candidate is TLD-shaped, return `candidates[0]` — the **shortest** `@`-bearing candidate. This is equivalent to "first hyphen after `@`" parsing, which is exactly the pre-fix behavior for any non-hyphenated email regardless of TLD length. So the parser is now strictly no worse than pre-fix for every user we previously handled correctly, plus it fixes the hyphenated-email cases that pre-fix broke.

Locked in with new test cases:
- `user@example.museum-some-app-name` → `{userId: "user@example.museum", packageName: "some-app-name"}`
- `user@example.industries-pkg-with-hyphens` → correct split (10-letter TLD)
- `user@127.0.0.1-pkg-with-hyphens` → correct split (IP-domain)

Two narrow cases still mis-handled and documented in `spec.md` as known limitations: (a) long-gTLD + hyphen-in-domain + no-live-session, (b) Punycode IDN where `.xn` falsely matches the TLD regex. Both resolve cleanly by migrating to the Public Suffix List (`tldts` npm package) — tracked as a follow-up so this fix can ship without scope creep.

124 tests pass, 0 fail.

**Comment by isaiahb** [comment]
@codex review

File: cloud/packages/cloud/src/services/websocket/bun-websocket.ts [hunk 3]
```diff
@@ -1032,27 +1033,22 @@ function isV3Sdk(version?: string): boolean {
...
-    return undefined;
-  }
-
-  const separator = sessionId.indexOf("-");
-  if (separator === -1 || separator === sessionId.length - 1) {
-    return undefined;
-  }
-
-  return sessionId.slice(separator + 1);
+  return parseLegacySessionId(sessionId, isActiveUserId)?.packageName;
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply cloud/packages/cloud/src/services/websocket/bun-websocket.ts:1036]: This is intentional behavior, codified by an existing test (`legacy-session-id.test.ts`: "longest match wins when multiple @-candidates are live").

The rationale: the lookup walks candidates longest-first because in any case where two userIds genuinely have live sessions and one is a strict hyphen-prefix of the other (the only way this fires), the longer userId is the more uniquely-identifying string and should win. Picking shortest would mean someone with the longer email gets routed to the wrong session.

For your specific example (`alpha@example.com-some-app` with `alpha@example.com-some` somehow live): this requires both users to actually exist with those exact email addresses AND both to be simultaneously connected via v2 SDK AND `alpha@example.com` to be the one sending the session ID. MongoDB enforces email uniqueness so the precondition needs `alpha@example.com-some` to be a real registered user — an email shape that's RFC-valid but extraordinarily uncommon in practice.

The package-extraction-shouldn't-depend-on-session-state observation has merit in the abstract, but the v3 SDK reconnect path (the only call site for `parsePackageNameFromLegacySessionId`) only hits the parser when `ws.data.packageName` AND JWT `packageName` are BOTH absent — which already represents a misconfigured v3 client. Adding a separate "parse-without-lookup" code path for that fallback would be more code surface for a scenario that's already an error case. Not worth it.

Resolving as not-actionable.
